servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/CapturedContextProviders.java

```diff
@@ -43,20 +41,32 @@ static List<CapturedContextProvider> providers() {
         return PROVIDERS;
     }
 
-    // TODO: this was copied from `ServiceLoaderUtils` because the `CapturedContextProvider` interface is package
-    //  private and the call to `ServiceLoader.load(..)` can only load implementations for types that are visible
-    //  to the package from which it is called (because it checks if it's accessible by checking stack trace).
-    //  One we make CapturedContextProvider public we can return to using ServiceLoaderUtils.
-    private static <T> List<T> loadProviders(final Class<T> clazz, final ClassLoader classLoader, final Logger logger) {
+    // This was copied from `ServiceLoaderUtils` because the implementation there logs the result, which normally
+    // wouldn't be a problem but as it turns out things like the MDC logging utils depend on the AsyncContext which
+    // may result in a cyclical dependency in class initialization.
+    private static <T> List<T> loadProviders(final Class<T> clazz, final ClassLoader classLoader) {
         final List<T> list = new ArrayList<>(0);
         for (T provider : ServiceLoader.load(clazz, classLoader)) {
             list.add(provider);
         }
         if (list.isEmpty()) {
-            logger.debug("ServiceLoader {}(s) registered: []", clazz.getSimpleName());
+            runLater(() -> getLogger().debug("ServiceLoader {}(s) registered: []", clazz.getSimpleName()));
```

Comment:
Is it worth checking that the log level is enabled before even trying to run a thread or you intentionally create the logger from a different thread?

Reply:
Even checking the log level can result in a cycle.

Reply:
It turns out doing that in DefaultAsyncContextProvider.java was the source of yet another issue (also now fixed).